### PR TITLE
[feat]: [PL-28352]: Added support to fetch hashicorp vault secret via path

### PR DESCRIPTION
**Pull request by nishantsaini20**
## Describe your changes
This PR is to add support to read hashicorp secrets without creating secret (reference secret) in harness. One can use the following expression to read secret present in hashicorp vault:
_<secret_manager_connector_scope>_**.hashicorpvault://**_<secret_manager_connector_identifier>_**/path/to/secret#key**
e.g. `org.hashicorpvault://my_vault/harness/testpath/test#value`
## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/37882)

### CODE REVIEW

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_Failed_** 
 


            
 #### 2 Issues 
* Code issue 2 Code issues 

 [Details](https://app.embold.io/organization/gh/harness/repositories/7752dbb3099d2dba3b5d37f5931a5e4c/pull-request/37882)

**Comment by nishantsaini20** [comment]
trigger messagecheck

**Comment by mohitg0795** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java` line 382 at r1](https://reviewable.io/reviews/harness/harness-core/37882#-ND2Jlmt121IpHCg5oZA:-ND2Jlmt121IpHCg5oZB:b-fevhsr) ([raw file](https://github.com/harness/harness-core/blob/23d40c633a8d0dbaf8df142ae51e4438b58f7dc4/120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java#L382)):*
> ```Java
>           String.format("Encryption type %s is not supported", encryptionTypeName), USER);
>     }
>     if (!dataRef.startsWith(SECRET_REFERENCE_DATA_ROOT_PREFIX)) {
> ```

what if dataRef is null?

**Comment by mohitg0795** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java` line 375 at r1](https://reviewable.io/reviews/harness/harness-core/37882#-ND2Jsdt0QrEM30p3haR:-ND2Jsdt0QrEM30p3haS:bvjdsg4) ([raw file](https://github.com/harness/harness-core/blob/23d40c633a8d0dbaf8df142ae51e4438b58f7dc4/120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java#L375)):*
> ```Java
>       encryptionTypeName = secretDetails[0];
>       dataRef = secretDetails[1];
>     } catch (ArrayIndexOutOfBoundsException exception) {
> ```

we should log the exception here otherwise how will we identify that this particular issue happened? I see same exception being thrown at other places as well.

**Comment by mohitg0795** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java` line 383 at r1](https://reviewable.io/reviews/harness/harness-core/37882#-ND2K15G7WP4nbpNeoYN:-ND2K15G7WP4nbpNeoYO:bgwa6ig) ([raw file](https://github.com/harness/harness-core/blob/23d40c633a8d0dbaf8df142ae51e4438b58f7dc4/120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java#L383)):*
> ```Java
>     }
>     if (!dataRef.startsWith(SECRET_REFERENCE_DATA_ROOT_PREFIX)) {
>       throw new SecretManagementException(INVALID_FORMAT, illegalFormatError, USER);
> ```

Shouldn't we be precise with the error here eg: We should mention that dataRef doesn't start with XYZ prefix as expected etc.

**Comment by mohitg0795** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java` line 394 at r1](https://reviewable.io/reviews/harness/harness-core/37882#-ND2KCsE62kVJwmjEi-g:-ND2KCsE62kVJwmjEi-h:b-9uorsu) ([raw file](https://github.com/harness/harness-core/blob/23d40c633a8d0dbaf8df142ae51e4438b58f7dc4/120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java#L394)):*
> ```Java
>           .relativePath(relativePath)
>           .build();
>     } catch (StringIndexOutOfBoundsException exception) {
> ```

log exception here

**Review by nishantsaini20 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37882)* status: 4 file(s) to review. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java` line 375 at r1](https://reviewable.io/reviews/harness/harness-core/37882#-ND2Jsdt0QrEM30p3haR:-ND2ivDy08IUWLppaEyk:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/23d40c633a8d0dbaf8df142ae51e4438b58f7dc4/120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java#L375)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

we should log the exception here otherwise how will we identify that this particular issue happened? I see same exception being thrown at other places as well.
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java` line 382 at r1](https://reviewable.io/reviews/harness/harness-core/37882#-ND2Jlmt121IpHCg5oZA:-ND2ixAu6Xvuzu5J4WL0:b8o20vt) ([raw file](https://github.com/harness/harness-core/blob/23d40c633a8d0dbaf8df142ae51e4438b58f7dc4/120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java#L382)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

what if dataRef is null?
</blockquote></details>

It can not be null. In case expression doesn't have delimiter ":" then initialisation of dataRef itself will throw  ArrayIndexOutOfBoundsException. So this line can not reach in this case. Am I missing any other case where this line can be reached when dataRef is null?
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java` line 383 at r1](https://reviewable.io/reviews/harness/harness-core/37882#-ND2K15G7WP4nbpNeoYN:-ND2jl5v8venxkWEtZhp:b-ps2jl8) ([raw file](https://github.com/harness/harness-core/blob/23d40c633a8d0dbaf8df142ae51e4438b58f7dc4/120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java#L383)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

Shouldn't we be precise with the error here eg: We should mention that dataRef doesn't start with XYZ prefix as expected etc.
</blockquote></details>

The main expression has many components in itself. It need detailed explanation. May be we can add url to the doc (once its available) as part of error message.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java` line 394 at r1](https://reviewable.io/reviews/harness/harness-core/37882#-ND2KCsE62kVJwmjEi-g:-ND2kCiYDpLZ4PGkQ9Hn:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/23d40c633a8d0dbaf8df142ae51e4438b58f7dc4/120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java#L394)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

log exception here
</blockquote></details>

Done.

File: 120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java [hunk 1]
```diff
@@ -338,6 +347,57 @@ public NGEncryptedData get(
...
+    return NGEncryptedData.builder()
+        .accountIdentifier(accountIdentifier)
+        .orgIdentifier(orgIdentifier)
+        .projectIdentifier(projectIdentifier)
+        .identifier(secretIdentifier)
+        .name(secretIdentifier)
+        .type(SettingVariableTypes.SECRET_TEXT)
+        .path(secretRefParsedData.getRelativePath())
+        .secretManagerIdentifier(secretRefParsedData.getSecretManagerIdentifier())
+        .encryptionType(EncryptionType.VAULT)
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java:363]: Should we add encryptionType in SecretRefParsedData so as to make it generic. Also we can set encryptionType during parsing in `validateAndGetSecretRefParsedData` ?

File: 120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java [hunk 2]
```diff
@@ -338,6 +347,57 @@ public NGEncryptedData get(
...
+    try {
+      encryptionTypeName = secretDetails[0];
+      dataRef = secretDetails[1];
+    } catch (ArrayIndexOutOfBoundsException exception) {
+      log.error(illegalFormatError, exception);
+      throw new SecretManagementException(INVALID_FORMAT, illegalFormatError, USER);
+    }
+    if (!EncryptionType.VAULT.getYamlName().equals(encryptionTypeName)) {
+      throw new SecretManagementException(UNSUPPORTED_OPERATION_EXCEPTION,
+          String.format("Encryption type %s is not supported", encryptionTypeName), USER);
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java:381]: should we add in error msg that it is not supported in expression evaluation or something related to direct access?

**Review by meenakshi-harness — CHANGES_REQUESTED** [review]

File: 970-ng-commons/src/main/java/io/harness/encryption/SecretRefParsedData.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+package io.harness.encryption;
+
+import lombok.Builder;
+import lombok.Getter;
+
+@Getter
+@Builder
+public class SecretRefParsedData {
+  public static final String SECRET_REFERENCE_EXPRESSION_DELIMITER = ":";
+  public static final String SECRET_REFERENCE_DATA_ROOT_PREFIX = "//";
```
Comment (meenakshi-harness) [inline 970-ng-commons/src/main/java/io/harness/encryption/SecretRefParsedData.java:17]: Just out of curiosity, why do we need `//` after `:` also as a separator in experssion `org.hashicorpvault://my_vault/harness/testpath/test#value` and why not `org.hashicorpvault:my_vault/harness/testpath/test#value`

**Review by nishantsaini20 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37882)* status: 4 file(s) to review. 1 change requests. 7. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java` line 363 at r2](https://reviewable.io/reviews/harness/harness-core/37882#-NCsGi_qQOqFVnY_0b2P-r2-363:-ND34RvJ9HQvpxYB3Xxg:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/a6b248d9e16654e3394708e71ca8e6796e713b9b/120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java#L363)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

Should we add encryptionType in SecretRefParsedData so as to make it generic. Also we can set encryptionType during parsing in `validateAndGetSecretRefParsedData` ?
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java` line 381 at r2](https://reviewable.io/reviews/harness/harness-core/37882#-NCsGi_qQOqFVnY_0b2P-r2-381:-ND34Zbt-DG9HqeMxhYn:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/a6b248d9e16654e3394708e71ca8e6796e713b9b/120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java#L381)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

should we add in error msg that it is not supported in expression evaluation or something related to direct access?
</blockquote></details>

Done.
___
*[`970-ng-commons/src/main/java/io/harness/encryption/SecretRefParsedData.java` line 17 at r2](https://reviewable.io/reviews/harness/harness-core/37882#-NCsGi_qQOqFVnY_0b2S-r2-17:-ND2rWH3729J_-VP9fNe:b-8h1flm) ([raw file](https://github.com/harness/harness-core/blob/a6b248d9e16654e3394708e71ca8e6796e713b9b/970-ng-commons/src/main/java/io/harness/encryption/SecretRefParsedData.java#L17)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

Just out of curiosity, why do we need `//` after `:` also as a separator in experssion `org.hashicorpvault://my_vault/harness/testpath/test#value` and why not `org.hashicorpvault:my_vault/harness/testpath/test#value`
</blockquote></details>

// after : To keep it similar to expression currently used in CG.

**Comment by mohitg0795** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java` line 382 at r1](https://reviewable.io/reviews/harness/harness-core/37882#-ND2Jlmt121IpHCg5oZA:-ND3wZ6WCSVtXgMF36cL:b-fdpfij) ([raw file](https://github.com/harness/harness-core/blob/23d40c633a8d0dbaf8df142ae51e4438b58f7dc4/120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java#L382)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

It can not be null. In case expression doesn't have delimiter ":" then initialisation of dataRef itself will throw  ArrayIndexOutOfBoundsException. So this line can not reach in this case. Am I missing any other case where this line can be reached when dataRef is null?
</blockquote></details>

Got it. I didn't go through the logic and commented on common sense of adding NULL check on all string comparisons. I agree here its not required, although its kinda good to have practice.

**Comment by mohitg0795** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java` line 394 at r1](https://reviewable.io/reviews/harness/harness-core/37882#-ND2KCsE62kVJwmjEi-g:-ND3x8DA7kCIWiO2g_3P:bxb02y1) ([raw file](https://github.com/harness/harness-core/blob/23d40c633a8d0dbaf8df142ae51e4438b58f7dc4/120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java#L394)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Approving for now, but I would suggest to have a unique logline for errors. Now, we have used "illegalFormatError" twice in this method itself. In case an exception happens, it would become difficult to conclude where did it fail.
Yes, you may argue that we can infer using the exception, but as a practice, I would suggest to avoid such things to make it clearer and not create cases of confusion that may arise due to such logs.

**Review by mohitg0795 — DISMISSED** [review]
> Reviewed 1 of 3 files at r3, 2 of 3 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37882)* status: 4 file(s) to review. 1 change requests. 3. unresolved discussions. 1 of 2 approvals obtained.

**Review by meenakshi-harness — APPROVED** [review]
> Reviewed 1 of 3 files at r3, 2 of 3 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37882)* status: :shipit: complete! ✓

**Review by mohitg0795 — APPROVED** [review]
> Reviewed 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37882)* status: 3 file(s) to review. 1 of 2 approvals obtained.

**Review by meenakshi-harness — APPROVED** [review]
> Reviewed 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37882)* status: :shipit: complete! ✓

**Pull request merged**